File: III_advanced/13_10_Trigraph.py
import numpy as np

count = 1
while True:
    n = int(input())
    if n == 0:
        break

    score = np.zeros((n, 3))
    for i in range(n):
        score[i] = list(map(int, input().split()))

    score[0, 0] = np.inf
    score[0, 2] += score[0, 1]

    for i in range(1, n):
        score[i, 0] += min(score[i-1, 0], score[i-1, 1])
        score[i, 1] += min(score[i-1, 0], score[i-1, 1], score[i-1, 2], score[i, 0])
        score[i, 2] += min(score[i-1, 1], score[i-1, 2], score[i, 1])

    print(f"{count}. {int(score[n-1, 1])}")
    count += 1

File: III_advanced/13_10_Trigraph.cpp
#include <iostream
#include <vector>
#include <algorithm>
#include <limits>

typedef std::vector<std::vector<int>> Vec2D;

int main() {
    int count = 1;

    while (true) {
        int n;
        std::cin >> n;
        if (n == 0) 
            break;
        

        Vec2D score(n, std::vector<int>(3));

        for (int i = 0; i < n; ++i) {
            std::cin >> score[i][0] >> score[i][1] >> score[i][2];
        }

        score[0][0] = std::numeric_limits<int>::max();
        score[0][2] += score[0][1];

        for (int i = 1; i < n; ++i) {
            score[i][0] += std::min(score[i-1][0], score[i-1][1]);
            score[i][1] += std::min({score[i-1][0], score[i-1][1], score[i-1][2], score[i][0]});
            score[i][2] += std::min({score[i-1][1], score[i-1][2], score[i][1]});
        }

        std::cout << count << ". " << score[n-1][1] << std::endl;
        count++;
    }

    return 0;
}
